gmp-tcs-offset/src/main/java/edu/gemini/aspen/gmp/tcsoffset/epics/EpicsTcsOffsetIOC.java

```diff
@@ -0,0 +1,563 @@
+            while ((inposOld != _tcsIsInPosition) && ( (System.currentTimeMillis()-t1) > FIVE_SECS )) {
+                inposOld = _tcsIsInPosition;
+                try {
+                    Thread.sleep(ONE_SEC);
+                } catch (InterruptedException e) {}
+                i++;
+            }
+            if ((System.currentTimeMillis()-t1) > FIVE_SECS ) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    
+    /**
+     * Execute the TCS apply action and wait until the TCS finish the action.
+     * The apply record execution is very fast, and I have never seen that the 
+     * 1000 milliseconds is consumed.
+     */
+
+    private void tcsApply() throws CAException, TimeoutException {
+        _tcsApply.setValue(Dir.START);
+        if (!waitTcs(ONE_SEC, CARSTATE.BUSY))
+            new TcsOffsetException(TcsOffsetException.Error.TIMEOUT,
+                    "TCS was not reached the BUSY state after applying the apply");
+        if (!waitTcs(ONE_SEC, CARSTATE.IDLE))
+            new TcsOffsetException(TcsOffsetException.Error.TIMEOUT,
+                    "TCS was not reached the IDLE state after applying the apply");
+    }
+
+
+    /**
+     * Set the TCS offset values in the TCS CA (<tcsTop>:poAdjust.X).
+     * @param val        :  P or Q value.
+     * @param offsetAngle:  90  -> p
+     *                      180 -> q
+     */
+
+    private void applyOffset(String val, String offsetAngle) throws CAException, TimeoutException {
+        _trackingFrameChannel.setValue(FRAME);
+        _offsetSizeChannel.setValue(val);
+        _angleChannel.setValue(offsetAngle);
+        _virtualTelChannel.setValue(VIRTUAL_TEL);
+        tcsApply();
+    }
+
+
+
+    /**
+     * This function manages the open or close loop sequence actions. The open and close sequence actions
+     * are defined in the configuration file. 
+     * @param loopKey: Indicates the loop sequence which will be executed. 
+     */
+    private void iterateSequence(String loopKey) throws CAException, TimeoutException, TcsOffsetException {
+        Iterator<String> keysLoop = _tcsChLoops.get(loopKey).getAsJsonObject().keySet().iterator();
+        LOG.fine("Starting the " + loopKey +" sequence.");
+        int indexCallFunc = -1;
+        while (keysLoop.hasNext()) {
+            String key = keysLoop.next();
+            String val = _tcsChLoops.get(loopKey).getAsJsonObject().get(key).toString().replace("\"","");
+            indexCallFunc = key.indexOf("$");
+            if (indexCallFunc == -1) {
+                if (_chLoops.get(key) != null)
+                    _chLoops.get(key).setValue(val);
+                else
+                    LOG.warning("The next command "+key+" can not be applied");
+            }
+            else {
+                String methodName = key.substring(indexCallFunc+1, key.length());
+                try {
+                    Method method = this.getClass().getDeclaredMethod(methodName);
```

Comment:
Yes, this is the case, because the configuration file defines the sequence commands to be executed to open and close the loops. There is a special action, which is the tcsApply. This action requires checking if the TCS could process the action after being executed, it is not just writing to the CA.